### DISCUSSION TITLE
Custom scalar type

### DISCUSSION THREAD

**Discussion by ghermaniov** [smallrye/smallrye-graphql#1164] [Q&A]
Hello guys,

I am currenty using smallrye graphql client with Qarkus, trying to call Shopify GraphQL API. I got stuck on this one because they are using for "returnUrl" a scalar of type "URL", this one: https://shopify.dev/api/admin-graphql/2021-10/scalars/URL.
Is there a way I can define this custom scalar myself?

I found this open issue on github and from what I've seen this is not possible yet: https://github.com/eclipse/microprofile-graphql/issues/94

Thanks.

Edit: I see both URL and URI from java are mapped to String in here: https://github.com/smallrye/smallrye-graphql/blob/main/common/schema-model/src/main/java/io/smallrye/graphql/schema/model/Scalars.java#L82
I guess this is the right way because most implementations will use a URL as String, maybe a way to re-map it to a custom defined one or a new annotation to specify a custom type that will be used when request is built, here: https://github.com/smallrye/smallrye-graphql/blob/cb91739b062d9bde6b59d46fe6f8a9da3165c8d8/client/implementation/src/main/java/io/smallrye/graphql/client/impl/typesafe/QueryBuilder.java#L58

**Comment by phillip-kruger** [comment]
@jmartisk @t1 ^^^ any suggestions ?

**Comment by jmartisk** [comment]
AFAIK the client library does not support adding custom mappings (for now), but it shouldn't be a problem to add a feature where it will automatically consider `URL` as a scalar and transform from a String to `URL`.  In your model classes you'd be able to use `URL` or `String` fields.
Would that be useful?

**Reply by ghermaniov** [reply]
Yes, definitely. Thanks.

**Reply by jmartisk** [reply]
I reported https://github.com/smallrye/smallrye-graphql/issues/1165 and will implement that soon :)

**Reply by jmartisk** [reply]
Ok I found a better way that is probably more suitable for your use case - turns out there is actually some support for custom scalars.
You need to create a custom scalar such as:

```
public class URL {

        private String string;

        public URL(String string) {
            this.string = string;
        }

        public String getString() {
            return string;
        }

        public static URL valueOf(String string) {
            return new URL(string);
        }
}
```

- the short name of the class should be the same as the name of the GraphQL type used by the server side
- the static `valueOf` method is important to be able to translate from a string to a  `URL` instance (and its presence makes the client library consider the class a scalar)

With a class like this, you can use it directly in your model classes.
I think this will be more suitable for your use case, because with automatic String-URL translating, it's impossible to tell the client library what GraphQL type is used to represent the URL on the server side. With this approach, you can control it by choosing the class name.

Will this work for you?

**Reply by t1** [reply]
There used to be some code that also would consider `java.net.URL` and `java.net.URI` to be valid scalar wrappers, given that they have a String-only constructor. But we then removed it due to #270. We could re-add it with the additional requirement that the class be either on a list of allowed wrapper classes (a built-in list containing, e.g., these two, plus a custom config option) or annotated accordingly (maybe reusing the server-side `@AdaptToScalar`).

**Reply by ghermaniov** [reply]
@jmartisk I tried that and It works like a charm. Just that I had to implement **toString()** method to return the internal **string**.
Thank you very much!
